From 85b044d15aca914fae2a784da14fa156a1030f44 Mon Sep 17 00:00:00 2001
From: md_5 <md_5@live.com.au>
Date: Sun, 3 Feb 2013 09:20:19 +1100
Subject: [PATCH] Handle Null Tile Entities


diff --git a/src/main/java/net/minecraft/world/level/World.java b/src/main/java/net/minecraft/world/level/World.java
index 9b786aff0..fa0391b5e 100644
--- a/src/main/java/net/minecraft/world/level/World.java
+++ b/src/main/java/net/minecraft/world/level/World.java
@@ -635,6 +635,13 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
 
         while (iterator.hasNext()) {
             TickingBlockEntity tickingblockentity = (TickingBlockEntity) iterator.next();
+            // Spigot start
+            if (tickingblockentity == null) {
+                getCraftServer().getLogger().severe("Spigot has detected a null entity and has removed it, preventing a crash");
+                iterator.remove();
+                continue;
+            }
+            // Spigot end
 
             if (tickingblockentity.isRemoved()) {
                 iterator.remove();
-- 
2.40.1

