From 33f5576fdf1f73522bd2c659cddd7ac8524888b7 Mon Sep 17 00:00:00 2001
From: DerFlash <bte@freenet.de>
Date: Tue, 9 Jul 2013 00:11:12 +0200
Subject: [PATCH] Save ticks lived to nbttag


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index c9f2ac3b6..dc52a9b8d 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1817,6 +1817,7 @@ public abstract class Entity implements INamableTileEntity, EntityAccess, IComma
             if (maxAirTicks != getDefaultMaxAirSupply()) {
                 nbttagcompound.putInt("Bukkit.MaxAirSupply", getMaxAirSupply());
             }
+            nbttagcompound.putInt("Spigot.ticksLived", this.tickCount);
             // CraftBukkit end
             IChatBaseComponent ichatbasecomponent = this.getCustomName();
 
@@ -1975,6 +1976,11 @@ public abstract class Entity implements INamableTileEntity, EntityAccess, IComma
             }
 
             // CraftBukkit start
+            // Spigot start
+            if (this instanceof EntityLiving) {
+                this.tickCount = nbttagcompound.getInt("Spigot.ticksLived");
+            }
+            // Spigot end
             this.persist = !nbttagcompound.contains("Bukkit.persist") || nbttagcompound.getBoolean("Bukkit.persist");
             this.visibleByDefault = !nbttagcompound.contains("Bukkit.visibleByDefault") || nbttagcompound.getBoolean("Bukkit.visibleByDefault");
             // SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
-- 
2.40.0

