From bda922b2bf84d4efa5bba872e2c4f5a1e7849597 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Mon, 20 Jan 2014 13:44:07 +1100
Subject: [PATCH] Catch stalling on corrupted map data / NBT arrays.


diff --git a/src/main/java/net/minecraft/nbt/NBTTagByteArray.java b/src/main/java/net/minecraft/nbt/NBTTagByteArray.java
index 4083563c6..2ab3a471e 100644
--- a/src/main/java/net/minecraft/nbt/NBTTagByteArray.java
+++ b/src/main/java/net/minecraft/nbt/NBTTagByteArray.java
@@ -16,6 +16,7 @@ public class NBTTagByteArray extends NBTList<NBTTagByte> {
         public NBTTagByteArray load(DataInput datainput, int i, NBTReadLimiter nbtreadlimiter) throws IOException {
             nbtreadlimiter.accountBytes(24L);
             int j = datainput.readInt();
+            com.google.common.base.Preconditions.checkArgument( j < 1 << 24); // Spigot
 
             nbtreadlimiter.accountBytes(1L * (long) j);
             byte[] abyte = new byte[j];
diff --git a/src/main/java/net/minecraft/nbt/NBTTagIntArray.java b/src/main/java/net/minecraft/nbt/NBTTagIntArray.java
index f0f4562d1..e5e90b9d7 100644
--- a/src/main/java/net/minecraft/nbt/NBTTagIntArray.java
+++ b/src/main/java/net/minecraft/nbt/NBTTagIntArray.java
@@ -16,6 +16,7 @@ public class NBTTagIntArray extends NBTList<NBTTagInt> {
         public NBTTagIntArray load(DataInput datainput, int i, NBTReadLimiter nbtreadlimiter) throws IOException {
             nbtreadlimiter.accountBytes(24L);
             int j = datainput.readInt();
+            com.google.common.base.Preconditions.checkArgument( j < 1 << 24); // Spigot
 
             nbtreadlimiter.accountBytes(4L * (long) j);
             int[] aint = new int[j];
-- 
2.40.1

